Q1: At very high resolution, the majority of the CpG nucleotides are unmethylated.

Unique sites in Nano: 53,346 (or 1.19% of all of the sites)

Unique sites in Bis: 132,466 (or 2.96% of all of the sites)

Q2: Bisulfite has a slightly higher average coverage than the nanopore, though the nanopore has a much more consistent coverage level when comapred to the bisulfite. 

Q3: The bisulfite method seems to be more effective at predciting methylation, though it may be overcalling methylation sites. 

Q4:Tumorigenesis results in a peak at about 50 at the violin plot for changes in methylation. This means that the tumor sample was about 50% more methylated than the normal, which could indicate that the tumor cells hijack the methylation machinery to write methylation marks on say for exmaple tumor suppressors which allows these cells to grow uncontrolled and block apoptosis. 

Q5:In this region in DNMT3A, there is a high methylation peak in the tumor sample that was not present in the normal sample. This could indicate that the tumor cells are adding methyl marks to the methyl transferase and blocking methylation of genes downstream of DNMT3A taht may block uncontrolled cell proliferation or induce apoptosis.  

Q6: Genetic imprinting is where one copy of a gene an individual inherits is silenced by methylation of the DNA. 

Q7: Phasing reads is determining whether two alleles (and chromosomes) are the same or not (homozygous or heterozygous). To phase reads, you need at least two heterozygous reads. There is a site where allele specfic methylation occurs, and in the tumor almost all of the reads have that spot as being methylated. This methylation is not consistent across all sites, as some are still mixed methylation even in the tumor reads. 

Q8: Not any set of reads can be phased. As mentioned there need to be two sequences of heterozygous individuals to phase reads. Phasing is also done primarily on long read sequencing as the long and accurate reads are unqiuely suited to phasing where short read sequences would not be. 

